File: front/lib/api/src/models/package.dart
import 'package:blee/api/src/models/image.dart';
import 'package:change_case/change_case.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'package.freezed.dart';
part 'package.g.dart';

@freezed
class Package with _$Package {
  const factory Package(
      {required String id,
      required String name,
      required String slug,
      @JsonKey(name: "release_year") required DateTime? releaseDate,
      @JsonKey(name: "artist_id") required String? artistId,
      @JsonKey(name: "artist_name") required String? artistName,
      required Image? poster}) = _Package;

  factory Package.fromJson(Map<String, dynamic> json) =>
      _$PackageFromJson(json);
}

enum PackageSort {
  name,
  addDate,
  artistName,
  releaseDate;

  @override
  String toString() {
    return this.name.toSnakeCase();
  }
}

File: front-vue/models/domain/package.ts
import * as yup from "yup";
import { Resource } from "./resource";
import { Image } from "./image";

const Package = Resource.concat(
    yup.object({
        name: yup.string().required(),
        slug: yup.string().required(),
        release_year: yup.date().required().nullable(),
        artist_id: yup.string().required().nullable(),
        artist_name: yup.string().required().nullable(),
        poster: Image.required().nullable(),
    }),
);

type Package = yup.InferType<typeof Package>;

type PackageSortingKeys = "name" | "addDate" | "artistName" | "release_date";

export { Package, type PackageSortingKeys };
